File: stack/stack.js

function Stack(/* optional */ capacity) {

  // Helper function to turn a negative
  // or undefined capcity into "no capacity"
  function cleanCapacity(capacity) {
    if (typeof capacity == "undefined") {
      return null;
    } else if (capacity < 0) {
      return null;
    } else {
      return capacity;
    }
  }

  // Private data used to manage this object
  let p = {
    data: [null],
    length: 0,
    capacity: cleanCapacity(capacity),
  }

  // Public API available
  let entity = {
    empty: () => p.length == 0,
    capacity: () => p.capacity === p.length,
    peek: () => p.data[p.length],
    push: (record) => {
      if (p.length === p.capacity) {
        return false;
      }

      p.length += 1;
      p.data[p.length] = record;
      return true;
    },
    pop: () => {
      const record = entity.peek();
      p.length = Math.max(p.length - 1, 0);
      return record;
    }
  };

  return entity;
}

module.exports = Stack;

File: stack/stack.py

class Stack:

  def __init__(self, max_capacity=None):
    self.data = []
    self.max_capacity = max_capacity

  def empty(self):
    return len(self.data) == 0;

  def capacity(self):
    if self.max_capacity is None:
      return False
    else:
      return len(self.data) >= self.max_capacity

  def push(self, val):
    if self.capacity():
      return False
    else:
      self.data.insert(0, val)
      return True

  def pop(self):
    if self.empty():
      return None
    else:
      val = self.data[0]
      del(self.data[0])
      return val

  def peek(self):
    if self.empty():
      return None
    else:
      return self.data[0]

